### ISSUE TITLE
Release v8.0

### ISSUE THREAD

**Issue by Shane32** [graphql-dotnet/graphql-dotnet#4004]
What features should be completed before releasing GraphQL.NET v8 ?

[See v8 milestone](https://github.com/graphql-dotnet/graphql-dotnet/milestone/30)

Completed: (among others)
- [x] Argument parsing occurs during validation
- [x] Validation rules can execute after arguments have been parsed
- [x] Complexity analyzer rewritten; can examine arguments to facilitate analysis
- [x] Federation support is simplified and Federation 2 is supported
- [x] Federation is supported for schema-first, code-first and type-first
- [x] ToObject is compiled
- [x] ComplexScalarGraphType added
- [x] `IMetadataWriter` and similar interfaces added to allow better intellisense
- [x] List converter enhancements and AOT support
- [x] Add OneOf Input Object support
- [x] Infer nullability by default for `Field` methods
- [x] Add Persisted Document support (draft spec)
- [x] Better error handling of GraphQL.NET pipeline; add timeout support
- [x] Optimize scalar lists
- [x] Allow interfaces to implement other interfaces

**Comment by gao-artur** [comment]
This one please! (server project) 
https://github.com/graphql-dotnet/server/issues/967

**Comment by Shane32** [comment]
I'm ready to release v8 any time now.  The remaining features that I would want to add seem to require a lot of work:

The following feature would be nice, as it's part of the official GraphQL specification:
- #1497 

Reworking the field builders so they only have applicable methods on them would be a big benefit, I think, but with the analyzers provided by @gao-artur there is much less benefit than previously.  See:
- #3583
- #2923 

There's some misc tasks related to type-first in #2893 that could be done... most are not breaking changes though.
- #2893 

There's some dataloader misc tasks that could be done.  Most are probably not breaking changes.  The biggest question is if we should move the data loader classes into the main GraphQL nuget package.  I don't know why they are separate; there are no additional dependencies in the dataloader package.

I'd really like to merge this code, but it needs work:
- #3550

So I don't really feel strongly motivated to work on any of the above tasks at the moment - lol !

**Comment by Shane32** [comment]
> This one please! (server project)
> 
> * [HttpStatus Code on GRAPHQL_VALIDATION error server#967](https://github.com/graphql-dotnet/server/issues/967)

Ok we should come up with suggested changes to the API

**Comment by gao-artur** [comment]
I may need to implement this analyzer
- https://github.com/graphql-dotnet/graphql-dotnet/issues/3823

depending on how much you want to implement this note from #3804
> We could set ALL undefined properties to their default values, not just init-only ones, making reflection, source-generated and dynamically-compiled versions simpler.

It looks like a breaking change, so if not now, we will have to wait for v9.
I have already tried implementing this analyzer, but `InputGraphTypeAnalyzer` is a big mess and will require refactoring to make the code extensible. With my limited free time now, it may take a long time.

**Comment by gao-artur** [comment]
There are a lot of obsolete API's we can remove now together with their analyzers and code fixes. I suggest adding a strong recommendation to the v8 migration notes to upgrade first to the latest v7 and apply all the automatic code fixes before upgrading to v8.

**Comment by Shane32** [comment]
> I suggest adding a strong recommendation to the v8 migration notes to upgrade first to the latest v7 and apply all the automatic code fixes before upgrading to v8.

Sure.  Typically I've released the last minor version alongside the new major version, marking anything deprecated in the prior version that we can in order to help prepare for the new major version.

> There are a lot of obsolete API's we can remove now together with their analyzers and code fixes.

I'm not sure there are any obsolete APIs that I would remove at this time.  I just skimmed through all uses of `[Obsolete]` and they are mostly what I expected.  I think we should consider marking `GraphQLMetadataAttribute` as obsolete.  To do so, I need to better understand its use, ensuring that alternative attributes are available.

**Comment by gao-artur** [comment]
How about `Field` methods that return `FieldType`? They were marked as obsolete in v5, and we have automatic code fixes to rewrite all of them with the `FieldBuilder` methods.

**Comment by Shane32** [comment]
> I may need to implement this analyzer [Ensure input object fields default don't conflict with source members default #3823](https://github.com/graphql-dotnet/graphql-dotnet/issues/3823) depending on how much you want to implement this note from #3804 

I'd skip for v8.  We don't have source generation yet anyway.

**Comment by Shane32** [comment]
> How about `Field` methods that return `FieldType`? They were marked as obsolete in v5, and we have automatic code fixes to rewrite all of them with the `FieldBuilder` methods.

So they were marked as obsolete for 2.5 years during all of v5 and v7?  Do the code fixes work if the method is removed?

I was sorta hoping that we would have an entire version cycle (v8) for users to utilize the automatic code fixes.  The analyzers have only existed since 7.7.0...

But it's been awhile as you say; we could remove them ...

**Comment by gao-artur** [comment]
Code fixes won't work because they need the analyzed code to compile. If there is something you still want to keep for v8, we can change the diagnostic level from warning to error (the user can change it back to warning with `.editorconfig`). This will nudge users to get rid of the obsolete code so it can be safely removed in v9.

**Comment by Shane32** [comment]
Good idea, although FYI I'm more thinking about users who only ever upgrade every few major versions.  We often were answering questions for people upgrading from v4 to v7, for example.  So based on how often we received issues of this nature, and my own team's coding patterns, I've been making a point to attempt to keep the bulk of the graph types' code stable (to the point at which it compiles) as long as possible, only requiring code changes to infrastructure that's changed.  (So I'm basically just talking about the `Field` methods.)  Now with the analyzers it should be much easier for people to perform the upgrade, but that doesn't mean that people are going to upgrade very often.  As the saying goes, if it ain't broke, don't fix it.

Maybe our rule of thumb going forward for `Field`-type methods is something like this:
- If there's an analyzer to replace the code, where version X is the version it was marked as obsolete:
  - Warning in Version X
  - Error in Version X+1
  - Removed in Version X+2

That would indicate we should remove the methods marked as obsolete in v5, which we can do if you want.

**Comment by Shane32** [comment]
Usually I just try to keep obsolete infrastructure code around for one major version, when reasonably feasible.

**Comment by Shane32** [comment]
Or provide some kind of patch/reversion for one major version, like the legacy complexity analyzer.

**Comment by gao-artur** [comment]
> We often were answering questions for people upgrading from v4 to v7

That's exactly why I suggested starting the migration notes from the recommendation to upgrade to the latest v7 first 🙂
I'd love to get rid of the obsolete code that sits there long enough.

**Comment by Shane32** [comment]
@gao-artur What's next for 8.0?  Are we ready to release?  The only thing I'd really want to do is #3550, which requires changes to the connection builders/types so they are AOT compatible, which I also want to do, but it's going to be a lot of changes, and probably mostly breaking changes, so I was going to leave it alone for v8.

**Comment by gao-artur** [comment]
I'd just want to remove obsolete apis and ensure UI projects have support for OneOf input types

**Comment by Shane32** [comment]
> I'd just want to remove obsolete apis 

- #4017

> ensure UI projects have support for OneOf input types

What are you thinking for this?  Ensure our UI packages are up to date?  It's still a proposed feature, not officially part of the spec, so they may not all support it yet.  That being said, it's a backwards-compatible change (essentially it's just a marker directive to indicate that the server may reject certain requests) so all UI projects should work nicely with it regardless.

**Comment by gao-artur** [comment]
I remember you said UI uses the graphql.js package, and the support for OneOf was [added](https://github.com/graphql/graphql-spec/pull/825#issuecomment-2238829753) in v16. I tried using the v7 version of Playground and GraphiQL, but they don't request `oneOf` fields from introspection. I tried to understand how the UI packed and how to upgrade it, but it looks like dark magic to me 😅

**Comment by Shane32** [comment]
GraphQL Playground has not been supported for a couple years now; I tested GraphiQL 3.5.0 (latest release) and it doesn't appear to support oneOf, and I looked through the Altair changelog and don't see support their either.  Do you want to look at this further?

**Comment by gao-artur** [comment]
No, it's ok. Let's releases it!

**Comment by gao-artur** [comment]
What about analyzers level for obsolete API? Do you want to change it to error?

**Comment by Shane32** [comment]
Personally, since `[Obsolete]` defaults to warning level, I'd leave the analyzers for obsolete code at warning level also.  Any obsolete methods still left in the code should execute correctly.  What do you think?  People can adjust the warning level pretty easily if they want to regardless.

**Comment by Shane32** [comment]
If you're done with analyzers for 8.0, then I'll work on releasing 7.9 and 8.0 and updating the other projects (server, etc) over the next week.  Not really sure how much time I'll have, so we'll see how it goes.  Server should be ready to go, at least.  I might do some organization on the migration doc if I have time also.

**Comment by gao-artur** [comment]
> Personally, since `[Obsolete]` defaults to warning level, I'd leave the analyzers for obsolete code at warning level also. Any obsolete methods still left in the code should execute correctly. What do you think? People can adjust the warning level pretty easily if they want to regardless.

Fine with me.

**Comment by Shane32** [comment]
Version 8 is released!  Probably need to update Conventions and the old Authorization/Relay packages also.

**Comment by gao-artur** [comment]
Great! Thank you!

**Comment by Shane32** [comment]
@gao-artur I'd be interested to know your upgrade experience with v8.  I've been pretty pleased on my end.  My helper libraries needed to be updated due to the changes to `IValidationRule` but pretty easy otherwise.

**Comment by gao-artur** [comment]
Didn't start yet. Hope to upgrade next week

**Comment by gao-artur** [comment]
I must admit this was the smoothest upgrade in the history of GraphQL.NET! Very few irreversible breaking changes and zero unexpected runtime errors.
Except two bugs (#4040 and #4042) I also encountered a false positive [GQL010](https://graphql-dotnet.github.io/docs/analyzers/gql010) diagnostic on the following input type
```c#
public class SystemFieldFilterGraphType : InputObjectGraphType<IList<SystemFieldExpressionParam>>
```
This graph type is generated from the DB configuration and has a registration in `ValueConverter` for `IList<SystemFieldExpressionParam>`. So, it worked well in runtime, but the static analyzer reported an error. I think this is a minor use case, and there is no need to fix the diagnostic. Users can suppress this diagnostic with `#pragma warning disable GQL010`, or, if they have many overrides in `ValueConverter`, set the severity to `none` in `.editorconfig`.

Also, moving the DataLoader's `ResolveAsync` extensions to `GraphQL` namespace didn't bring the expected benefit. Because the `GraphQL` namespace is usually not imported, these extension methods still have poor discoverability. The only using that is certainly imported is `GraphQL.Types`, where the `ObjectGraphType`,  `InputObjectGraphType`, and `InterfaceGraphType` are defined. Maybe we should consider moving them into `GraphQL.Types` in the next major version.

The only thing I would like to ask before releasing 8.0.2 is adding argument type validation in `PatternMatchingVisitor` because currently, it silently ignores the directive if the type is not `string.` I thing it should throw an exception.

**Comment by Shane32** [comment]
> The only thing I would like to ask before releasing 8.0.2 is adding argument type validation in `PatternMatchingVisitor` because currently, it silently ignores the directive if the type is not `string.` I thing it should throw an exception.

Seems reasonable. Want to create a PR?

**Comment by gao-artur** [comment]
Sure

**Comment by gao-artur** [comment]
- #4044
